ath79: add support for Netgear EX6400 and EX7300

danielg4 opened the pull request:
This is sold as a dual-band 802.11ac range extender. It has a sliding
switch for Extender mode or Access Point mode, a WPS button, a recessed
Reset button, a hard-power button, and a multitude of LED's, some
multiplexed via an NXP 74AHC164D chip. The internal serial header pinout is
Vcc, Tx, Rx, GND, with GND closest to the corner of the board. You may
connect at 115200 bps, 8 data bits, no parity, 1 stop bit.

Specification:
- System-On-Chip: QCA9558
- CPU/Speed: 720 MHz
- Flash-Chip: Winbond 25Q128FVSG
- Flash size: 16 MiB
- RAM: 128 MiB
- Wireless No1: QCA9558 on-chip 2.4GHz 802.11bgn, 3x3
- Wireless No2: QCA99x0 chip 5GHz 802.11an+ac, 4x4
- PHY: Atheros AR8035-A

Installation:
If you can get to the stock firmware's firmware upgrade option, just feed
it the factory.img and boot as usual. As an alternative, TFTP the
factory.img to the bootloader.

Signed-off-by: Daniel Gimpelevich <daniel@gimpelevich.san-francisco.ca.us>

danielg4 committed:
ath79: add support for Netgear EX6400 and EX7300

This is sold as a dual-band 802.11ac range extender. It has a sliding
switch for Extender mode or Access Point mode, a WPS button, a recessed
Reset button, a hard-power button, and a multitude of LED's, some
multiplexed via an NXP 74AHC164D chip. The internal serial header pinout is
Vcc, Tx, Rx, GND, with GND closest to the corner of the board. You may
connect at 115200 bps, 8 data bits, no parity, 1 stop bit.

Specification:
- System-On-Chip: QCA9558
- CPU/Speed: 720 MHz
- Flash-Chip: Winbond 25Q128FVSG
- Flash size: 16 MiB
- RAM: 128 MiB
- Wireless No1: QCA9558 on-chip 2.4GHz 802.11bgn, 3x3
- Wireless No2: QCA99x0 chip 5GHz 802.11an+ac, 4x4
- PHY: Atheros AR8035-A

Installation:
If you can get to the stock firmware's firmware upgrade option, just feed
it the factory.img and boot as usual. As an alternative, TFTP the
factory.img to the bootloader.

Signed-off-by: Daniel Gimpelevich <daniel@gimpelevich.san-francisco.ca.us>

chunkeey commented:
@danielg4 I'm new(-ish) here and I'm looking for candidates to merge. Are you still interested in adding support for these devices? if yes, It would be great if you could split the EX6400 and EX7300 into separate devices even though they are internally identical. Thanks.

danielg4 commented:
It'll be at least a week before I can get to it, but yes, I'm interested. I will split this PR into two commits with a push to resolve the merge conflict, but splitting it into separate devices makes zero sense, because they are _firmware_-identical, not necessarily internally identical. There is still an unresolved issue of the wifi signal being much stronger with the factory firmware than with OpenWrt, and I'm not yet sure why, but if I can fix that, I'd prefer to do it before merge.

rmilecki commented on target/linux/ath79/dts/qca9558_netgear_ex7300.dts:
```diff
@@ -0,0 +1,233 @@
+				reg = <0x080000 0x010000>;
+			};
+
+			partition@90000 {
```
Please add `compatible`, see
[OpenWrt-Devel] Specifying "firmware" partition format on all DTS targets
http://lists.infradead.org/pipermail/openwrt-devel/2018-November/014706.html
for details.

danielg4 committed:
ath79: remove more duplicate image build variables

Remove Netgear-specific image build variables which are set to the same
value.

Signed-off-by: Daniel Gimpelevich <daniel@gimpelevich.san-francisco.ca.us>

danielg4 commented:
OK, this force-push is so far completely **UNTESTED** and not ready for merge, but I'm requesting feedback on it prior to continuing. Thanks.

ynezz commented on target/linux/ath79/dts/qca9558_netgear_ex7300.dts:
```diff
@@ -0,0 +1,234 @@
+		power_amber: power_amber {
+			label = "netgear:amber:power";
+			gpios = <&gpio 20 GPIO_ACTIVE_LOW>;
+			default-state = "off";
```
From https://openwrt.org/submitting-patches#dts_checklist
> Remove all ocurrencies of `default-state = “off”` properties under your LED nodes ([details](https://github.com/openwrt/openwrt/commit/0d23fd2ab29a66f5d03187db4fac3e396b4f3b62))

ynezz commented on target/linux/ath79/image/generic.mk:
```diff
@@ -421,6 +421,20 @@ define Device/netgear_wndr3x00
+  NETGEAR_BOARD_ID := EX7300series
+  NETGEAR_HW_ID := 29765104+16+0+128
+  IMAGE_SIZE := 15552k
+  SUPPORTED_DEVICES += ex7300
```
You can drop this, it's being added automatically for you in https://github.com/openwrt/openwrt/blob/eb1887be93a765b4b86ecad14dd72a5a72d6870e/target/linux/ath79/image/Makefile#L63

danielg4 replied:
Uh-oh. This may break what I did in the other commit, as well as the existing "wndr3x00" entry.

danielg4 replied:
On second look, I'm not sure you are correct, and the other Netgear devices all have such a line.

mkresin replied:
He is correct, while his explanation is misleading. The correct supported devices entry for your board is added by the mentioned code line, to ensure sysupgrade accepts the image.

The extension of the SUPPORTED_DEVICES is only required if the board used a different userspace boardname in the past and you want to upgrade from an image using the old name to this image (which is using the new name).

Since you add support for a new board, you don't need to take care of former used userspace boardnames.

danielg4 replied:
In that case, is what I did in the f55a944 commit incorrect?

ynezz commented on target/linux/ath79/dts/qca9558_netgear_ex7300.dts:
```diff
@@ -0,0 +1,234 @@
+	phy-handle = <&phy4>;
+	phy-mode = "rgmii";
+
+	pll-data = <0x86000000 0x80000101 0x80001313>;
```
Where did you got this values from?

danielg4 replied:
I derived them from the vendor source tarball and traced through the driver code.

danielg4 commented on target/linux/ath79/dts/qca9558_netgear_ex7300.dts:
```diff
@@ -0,0 +1,234 @@
+
+			uboot: partition@0 {
+				label = "u-boot";
+				reg = <0x000000 0x040000>;
```
I did it not only for column alignment, but also to be consistent with other .dts files in OpenWrt.

rmilecki replied:
Seems like a common practice, a matter of taste, I'm OK with that.

danielg4 commented on target/linux/ath79/dts/qca9558_netgear_ex7300.dts:
```diff
@@ -0,0 +1,234 @@
+			gpios = <&gpio 19 GPIO_ACTIVE_LOW>;
+		};
+
+		power_amber: power_amber {
```
I left this in case someone has a better idea for the aliases prior to merge, or perhaps after.

chunkeey commented on target/linux/ath79/dts/qca9558_netgear_ex7300.dts:
```diff
@@ -0,0 +1,234 @@
+			partition@90000 {
+				label = "firmware";
+				reg = <0x090000 0xf30000>;
+				compatible = "denx,uimage";
```
Shouldn't this be "netgear,uimage" ?

chunkeey commented on target/linux/ath79/image/generic.mk:
```diff
@@ -421,6 +421,20 @@ define Device/netgear_wndr3x00
+  NETGEAR_HW_ID := 29765104+16+0+128
+  IMAGE_SIZE := 15552k
+  SUPPORTED_DEVICES += ex7300
+  DEVICE_PACKAGES := kmod-ath10k ath10k-firmware-qca99x0
```
61b5b4971e7d6aab6b0695997c3b5aaf73c53b5f made ath10k-ct the default.

kmod-ath10k-ct ath10k-firmware-qca99x0-ct

danielg4 replied:
Oops. Will fix.

danielg4 commented:
Should be mergeable now.

danielg4 commented:
Still waiting for @chunkeey or other ACK/NAK...

rmilecki commented:
Looks OK assuming `compatible = "denx,uimage";` was verified to work.

danielg4 commented:
> Looks OK assuming `compatible = "denx,uimage";` was verified to work.

The image works with that, and the code looks like it should, but I'm not sure it's actually used currently.

chunkeey commented:
> Still waiting for @chunkeey or other ACK/NAK...

See https://github.com/openwrt/openwrt/pull/1379#issuecomment-447923069

> >There is still an unresolved issue of the wifi signal being much stronger with the factory firmware than with OpenWrt, and I'm not yet sure why, but if I can fix that, I'd prefer to do it before merge.

danielg4 commented:
> > > There is still an unresolved issue of the wifi signal being much stronger with the factory firmware than with OpenWrt, and I'm not yet sure why, but if I can fix that, I'd prefer to do it before merge.

I was unable to duplicate the discrepancy with the latest factory firmware and the latest builds, so I'd say this is tentatively resolved.

danielg4 commented:
Still waiting for merge.

danielg4 commented:
I was referred to target/linux/ath79/dts/ar9331_tplink_tl-wr740n-v4.dts and target/linux/ath79/dts/ar9331_tplink_tl-wr741nd-v4.dts as examples, and I only now realize that they were to be examples of the same .dts to be used for different devices. Doing that here would lead to two separate images being generated, with different board signatures, and I'm not so sure that's a good idea. Netgear also uses the same image file for both devices.

danielg4 commented:
In the case of the example devices, the TPLINK_HWID is different, requiring separate images, which is not the same situation.

danielg4 commented:
I do have a question before merge, though: In some of the qca9558_*.dts files, the compatible string reads "qca,qca9558" but others have "qca,qca9557" instead. Which is correct?

danielg4 commented:
Also requesting guidance on which of the LED aliases in the .dts, if any, should be switched to power_amber.

ynezz commented:
> It would be great if you could split the EX6400 and EX7300 into separate devices even though they are internally identical.

I can't find a valid reason why the split is no longer necessary.

> Doing that here would lead to two separate images being generated, with different board signatures, and I'm not so sure that's a good idea

It was requested for a reason. I wouldn't flash image named EX7300 to device with EX6400 on the sticker. And I barely remember similar confusing situation with other device/image just a few months ago.

> should be switched to power_amber.

My personal taste would be `led-failsafe` and `led-upgrade`.

danielg4 commented:
> I do have a question before merge, though: In some of the qca9558_*.dts files, the compatible string reads "qca,qca9558" but others have "qca,qca9557" instead. Which is correct?

This question remains.

ynezz commented on target/linux/ath79/dts/qca9558_netgear_ex7300.dtsi:
```diff
@@ -0,0 +1,223 @@
+
+			partition@70000 {
+				label = "config";
+				reg = <0x070000 0x010000>;
```
Perhaps `read-only`?

danielg4 replied:
This partition stores the current settings for the factory firmware. There is no need for it to be read-only.

ynezz replied:
> This partition stores the current settings for the factory firmware. There is no need for it to be read-only.

So what happens if you rewrite it accidentally with some garbage?

danielg4 replied:
On OpenWrt, nothing. On the factory firmware, undefined, but cured with a factory reset.

ynezz commented on target/linux/ath79/dts/qca9558_netgear_ex7300.dtsi:
```diff
@@ -0,0 +1,223 @@
+
+			partition@80000 {
+				label = "pot";
+				reg = <0x080000 0x010000>;
```
Perhaps `read-only`?

danielg4 replied:
This is a timestamp partition. No need for it to be read-only.

ynezz replied:
> No need for it to be read-only.

I would rather say, that there's no need to have it writtable or is it?

danielg4 replied:
Upon reverting to factory firmware, it may be useful to store some value in here, not sure.

ynezz commented on target/linux/ath79/dts/qca9558_netgear_ex7300.dtsi:
```diff
@@ -0,0 +1,223 @@
+
+			partition@fc0000 {
+				label = "language";
+				reg = <0xfc0000 0x040000>;
```
Perhaps `read-only`?

danielg4 replied:
Maybe. Rationale?

ynezz replied:
To prevent accidental overwrite of the data, making reverting back to stock/factory firmware PITA.

danielg4 replied:
I thought of that, but then I also thought of needing to overwrite it with what the factory firmware provides. However, it appears that the factory firmware might not provide it, and on my device (which had a mostly-erased flash chip when I got it) this partition (and u-boot-env) look rather blank with no apparent ill effects on the factory firmware.

ynezz commented:
> I do have a question before merge, though: In some of the qca9558_*.dts files, the compatible string reads "qca,qca9558" but others have "qca,qca9557" instead. Which is correct?

I would say that it should be set to the actual SoC as used on the device. I don't know the real reason behind this inconsistent usage in the current tree (maybe just another copy&paste disease?)

danielg4 commented:
> > I do have a question before merge, though: In some of the qca9558_*.dts files, the compatible string reads "qca,qca9558" but others have "qca,qca9557" instead. Which is correct?
> 
> I would say that it should be set to the actual SoC as used on the device. I don't know the real reason behind this inconsistent usage in the current tree (maybe just another copy&paste disease?)

At the time that I started the ath79 branch of this port, they all said "qca,qca9557" and all the ones that say "qca,qca9558" have been added/edited since.

ynezz commented:
I've merged it to my staging tree https://git.openwrt.org/?p=openwrt/staging/ynezz.git
I've fixed some whitespace issues not visible at first sight here on Github and reordered few vars in image makefile:

```diff
diff --git a/target/linux/ath79/dts/qca9558_netgear_ex7300.dtsi b/target/linux/ath79/dts/qca9558_netgear_ex7300.dtsi
index 6ccc4ad..21c25a5 100644
--- a/target/linux/ath79/dts/qca9558_netgear_ex7300.dtsi
+++ b/target/linux/ath79/dts/qca9558_netgear_ex7300.dtsi
@@ -92,7 +92,6 @@
 			label = "netgear:green:router";
 			gpios = <&led_gpio 0 GPIO_ACTIVE_LOW>;
 		};
-
 	};
 
 	keys {
diff --git a/target/linux/ath79/image/common-netgear.mk b/target/linux/ath79/image/common-netgear.mk
index b8664c0..9b79936 100644
--- a/target/linux/ath79/image/common-netgear.mk
+++ b/target/linux/ath79/image/common-netgear.mk
@@ -39,4 +39,3 @@ define Device/netgear_ath79
   IMAGE/sysupgrade.bin := $$(IMAGE/default) | append-metadata | check-size $$$$(IMAGE_SIZE)
   IMAGE/factory.img := $$(IMAGE/default) | netgear-dni | check-size $$$$(IMAGE_SIZE)
 endef
-
diff --git a/target/linux/ath79/image/generic.mk b/target/linux/ath79/image/generic.mk
index 2717ad8..241c1e2 100644
--- a/target/linux/ath79/image/generic.mk
+++ b/target/linux/ath79/image/generic.mk
@@ -547,19 +547,19 @@ TARGET_DEVICES += pcs_cr5000
 define Device/netgear_wndr3x00
   ATH_SOC := ar7161
   IMAGE/default := append-kernel | pad-to $$$$(BLOCKSIZE) | netgear-squashfs | append-rootfs | pad-rootfs
-  $(Device/netgear_ath79)
   DEVICE_PACKAGES := kmod-usb-core kmod-usb-ohci kmod-usb2 kmod-usb-ledtrig-usbport kmod-leds-reset kmod-owl-loader
+  $(Device/netgear_ath79)
 endef
 
 define Device/netgear_ex7300_ex6400
   ATH_SOC := qca9558
-  IMAGE/default := append-kernel | pad-offset $$$$(BLOCKSIZE) 64 | netgear-rootfs | pad-rootfs
-  $(Device/netgear_ath79)
   NETGEAR_KERNEL_MAGIC := 0x27051956
   NETGEAR_BOARD_ID := EX7300series
   NETGEAR_HW_ID := 29765104+16+0+128
   IMAGE_SIZE := 15552k
+  IMAGE/default := append-kernel | pad-offset $$$$(BLOCKSIZE) 64 | netgear-rootfs | pad-rootfs
   DEVICE_PACKAGES := kmod-ath10k-ct ath10k-firmware-qca99x0-ct
+  $(Device/netgear_ath79)
 endef
 
 define Device/netgear_ex6400
```

So if you agree with this changes, please close this PR so we can move on. Thanks.

danielg4 commented:
ACK.

adrianschmutzler commented:
Although this PR has been has been closed already, I would be interested in some additional data about the device(s) in case you still have access to it:

I'm currently working on a PR to store the label-mac-address in the DTS (https://github.com/openwrt/openwrt/pull/2159).

I would appreciate if you could provide information about your device:
Please run
`find /proc/device-tree/ -name "*mac-address*"`
on your device (just with the current firmware, no need to update anything; just has to be ath79, no ar71xx).
For each of the returned paths (if there are any), retrieve the mac-address with `cat`, e.g.
`cat /proc/device-tree/ahb/eth@19000000/mac-address`

Now check which of those returned MAC addresses matches the one on the label/sticker/case of your device and tell me the device-tree path leading to it. Note that you do not have to post your MAC addresses here, but just a statement about which match, e.g. like
```
/proc/device-tree/ahb/eth@1a000000/mac-address
/proc/device-tree/ahb/eth@1a000000/mtd-mac-address-increment
/proc/device-tree/ahb/eth@1a000000/mtd-mac-address
/proc/device-tree/ahb/eth@19000000/mac-address <== this one
/proc/device-tree/ahb/eth@19000000/mtd-mac-address
/proc/device-tree/ahb/apb/wmac@18100000/mac-address
/proc/device-tree/ahb/apb/wmac@18100000/mtd-mac-address
```

There may be one, two or no paths giving the correct address. It may also happen that the `find` command returns nothing. Thanks.

adrianschmutzler commented:
Unfortunately, my description is slightly incorrect. The MAC addresses are stored binary, so you can't read with `cat`, but have to use the following code to read from the paths retrieved by `find`:
```
. /lib/functions/system.sh
get_mac_binary "$yourpath" 0
```
The include is obviously only required once. Sorry.

danielg4 commented:
```
root@OpenWrt:/# find /proc/device-tree/ -name "*mac-address*"
/proc/device-tree/ahb/eth@19000000/mac-address
/proc/device-tree/ahb/eth@19000000/mtd-mac-address
/proc/device-tree/ahb/apb/wmac@18100000/mac-address
/proc/device-tree/ahb/apb/wmac@18100000/mtd-mac-address
```

The label has four fields with bar codes: S/N, MAC 1, MAC 2, and WPS PIN. Both of the "mac-address" nodes (the first and the third above) match MAC 1. The other two are eight bytes each and not MAC addresses at all.

adrianschmutzler commented:
> Both of the "mac-address" nodes (the first and the third above) match MAC 1.

Thanks. Are you really sure?

In DTS, eth0 and wmac have different locations, so I would expect them to be different:
```
&wmac {
	mtd-mac-address = <&caldata 0x06>;
};
&eth0 {
	mtd-mac-address = <&caldata 0x00>;
};
```

danielg4 commented:
Yes, eth0 and wmac have the same MAC address, which is located twice in the flash, so I used both locations in the DTS. The second radio (pcie0) has the other MAC address.

adrianschmutzler commented:
Okay, interesting. So I will use `label-mac-device = &eth0;`.

danielg4 commented:
Both MAC addresses are printed on the label separately, MAC 1 (eth0/wmac) and MAC 2 (pcie0).

danielg4 commented:
The flash partition begins with three MAC addresses back-to-back, the first two of which happen to be the same.

adrianschmutzler commented:
label-mac-device is meant to derive a single MAC address which is used to identify the device. This may be used to e.g. label a device in a network with an address that is obvious to the user (= on the label).
Thus, within the boundaries of this concept, I choose MAC 1 .
